### PR TITLE
Update the documentation guidelines

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

Pulling some documentation aspects proposed in #10225 into the documentation guidelines.

The guidelines do still need further updates. However, lets start with these so that the amount of changes is still manageable for reviewers.

### CODE REVIEW

**Comment by fredrik-1** [comment]
I agree with those changes. One thing I have though about though and I was about to write about in the discussion. I have no problem if this get merge directly though. 

I believe that the matplotlib in for example `````~matplotlib.axes.Axes`````could be left out most of the times. It makes the string longer without given much information for most users. 

I believe that it often make sense to include the module name also in the rendered documentation. Some docstrings, like for the colorbar module, looks strange when it is rendered as `colorbar` is a thin wrapper  of `colorbar`. It would be easer to read as `pyplot.colorbar` is a thin wrapper  of `Figure.colorbar`.

I believe that I in most cases want to read the module name also in the rendered documentation.

File: doc/devel/documenting_mpl.rst [hunk 1]
```diff
@@ -352,104 +379,143 @@ See the `~.Axes.hlines` documentation for how this renders.
...
+      norm : `~matplotlib.colors.Normalize`, optional
+         A `.Normalize` instance is used to scale luminance data to 0, 1.
 
-nor the ````literal```` role:
+   .. note::
 
-.. code-block:: rst
+      Sphinx claims to automatically link types in ``Parameters`` and
+      ``Returns`` sections. However, this does currently not work reliably.
+      Therefore we still have to enclose them in backticks.
```
Comment (fredrik-1) [inline doc/devel/documenting_mpl.rst]: Is this true? I believe that the only section where backticks are not necessary is the `See Also` section.

**Comment by NelleV** [comment]
That looks good to me (although I am worried about the length of the document. Those are typically never read by our contributors, including core-contributors)

**Comment by timhoffm** [comment]
Even if not everyone follows those guidelines exactly, it's still of value. It can be seen as an aspirational goal, and if we follow at least partly we're still converging in that direction. Actually, there have been requests for a better documentation guide. After all, it can help a contributor to find out how to do certain things. We have a lot of different styles in the actual docs, because the guidelines have changed over time. Having an up-to-date document to reference when in doubt is a good thing. This is essentially a collection of proven best practices.

The length can actually be a bit daunting. I've tried to cut back on duplication, give it more structure and add examples, so that one can find out certain aspects without having to read the complete document. It's still far from perfect. There can be done more to clean up the guidelines, but it's a step in the right direction.

**Comment by NelleV** [comment]
If we don't get another review on this soon, I'll just go ahead and merge it.

File: doc/devel/documenting_mpl.rst [hunk 2]
```diff
@@ -213,38 +220,52 @@ is better than:
...
-  `~matplotlib.collections.LineCollection`
+  `matplotlib.collections.LineCollection`
 
-returns a link to the documentation of
-`~matplotlib.collections.LineCollection`.  For the full path of the
-class to be shown, omit the tilde:
+generates a link like this: `matplotlib.collections.LineCollection`.
 
-.. code-block:: python
+*Note:* We use the sphinx setting `default_role = 'obj'` so that you don't
```
Comment (choldgraf) [inline doc/devel/documenting_mpl.rst]: why not make these `.. note::` so it'll render that way in the site?

File: doc/devel/documenting_mpl.rst [hunk 3]
```diff
@@ -213,38 +220,52 @@ is better than:
...
+have to use qualifiers like `:class:`, `:func:`, `:meth:` and the likes.
```
Comment (choldgraf) [inline doc/devel/documenting_mpl.rst]: On that note, I believe that these need to be double-backticks in rST, no?

Comment (NelleV) [reply doc/devel/documenting_mpl.rst]: I thought this as well, but I believe we are using the very nice option that makes the documentation build fail if there's any warnings, so I guess these are correct.

Comment (choldgraf) [reply doc/devel/documenting_mpl.rst]: it may not give a warning though, it may just not display as in-line code, but as regular text w/ backticks

Comment (timhoffm) [reply doc/devel/documenting_mpl.rst]: You're right. These should be double-backticks. Fixed.

File: doc/devel/documenting_mpl.rst (hunk 2 above)
Comment (timhoffm) [reply doc/devel/documenting_mpl.rst]: This is rather a side note. It's not something you have to know to be able to write documentation. As such a `.. note::` admonition would give it too much attention.

**Comment by NelleV** [comment]
Ok, this looks great. Thanks @choldgraf for the review (despite the bad timing :smile: ) and thanks @timhoffm for updating the docs on writing docs!

**Comment by Kojoley** [comment]
Backporting this to v2.2.x to fix CI issues there

@meeseeksdev backport to v2.2.x

**Pull request merged**